
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta name="google-site-verification" content="ptuXe7ex0jPvowLOm_BRWnxrV-EfJeXDTA7zF0aLuzk" />
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

<!-- Begin Jekyll SEO tag v2.8.0 -->
<title>AptaScreen | A script to screen Aptamer sequences enriched in library</title>
<meta name="generator" content="Jekyll v3.9.3" />
<meta property="og:title" content="AptaScreen" />
<meta property="og:locale" content="en_US" />
<meta name="description" content="A script to screen Aptamer sequences enriched in library" />
<meta property="og:description" content="A script to screen Aptamer sequences enriched in library" />
<link rel="canonical" href="https://fengjiang94.github.io/AptaScreen/" />
<meta property="og:url" content="https://fengjiang94.github.io/AptaScreen/" />
<meta property="og:site_name" content="AptaScreen" />
<meta property="og:type" content="website" />
<meta name="twitter:card" content="summary" />
<meta property="twitter:title" content="AptaScreen" />
<script type="application/ld+json">
{"@context":"https://schema.org","@type":"WebSite","description":"A script to screen Aptamer sequences enriched in library","headline":"AptaScreen","name":"AptaScreen","url":"https://fengjiang94.github.io/AptaScreen/"}</script>
<!-- End Jekyll SEO tag -->

    <link rel="stylesheet" href="/AptaScreen/assets/css/style.css?v=f0fb77ff6009eba71f3710c11b4dc99469d7e584">
    <!-- start custom head snippets, customize with your own _includes/head-custom.html file -->

<!-- Setup Google Analytics -->



<!-- You can set your favicon here -->
<!-- link rel="shortcut icon" type="image/x-icon" href="/AptaScreen/favicon.ico" -->

<!-- end custom head snippets -->

  </head>
  <body>
    <div class="container-lg px-3 my-5 markdown-body">
      

      <h1 id="aptascreen">AptaScreen</h1>
<p>A script to screen Aptamer sequences enriched in library following SELEX (Systematic Evolution of Ligands by EXponential Enrichment)</p>

<p>If you use this pipeline, please consider citing it:</p>

<ol>
  <li>Jiang, F., Hedaya, O.M., Khor, E., Wu, J., Auguste, M., &amp; Yao, P. (2023). <a href="https://academic.oup.com/nar/article/51/11/5831/7147500">RNA binding protein PRRC2B mediates translation of specific mRNAs and regulates cell cycle progression</a>. <em>Nucleic Acids Research</em>, 51(11), 5831–5846</li>
</ol>

<h2 id="software-prerequisites">Software prerequisites</h2>
<p>This pipeline is designed to run on Linux servers, and requires the following software:</p>

<p>They need to be installed and added to the $PATH before using the pipeline.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>R
fastx
cutadapt
</code></pre></div></div>

<p>In addition, the following R packages are also required. please install them.</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pillar
dplyr
data.table
ggplot2
tidyr
</code></pre></div></div>

<h2 id="pipeline-setup">Pipeline setup</h2>

<p>1, Download scripts from github to Linux server:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>git clone https://github.com/fengjiang94/AptaScreen.git
</code></pre></div></div>

<p><strong>AptaScreen.sh</strong> is the main script to run.</p>

<ol>
  <li>add AptaScreen to $path
    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>export PATH=$PATH:/Where AptaScreen is/AptaScreen
</code></pre></div>    </div>
  </li>
</ol>

<p>3, Add executable permissions</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>chmod +x /Where AptaScreen is/AptaScreen/AptaScreen.sh
</code></pre></div></div>

<p>4, cutadapt and dplyr</p>

<p>Occasionally cutadapt and dplyr are not compatible so that we unload cutadapt before loading dplyr in line 292 of <strong>AptaScreen.sh</strong></p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>module unload cutadapt # we have to unload cutadapt since it is not compatible with r-dplyr, whcih is needed for the following steps. This may be a problem with our local cluster only.
</code></pre></div></div>

<p>If you are not encountering this problem or not loading cutadapt as Environment Modules, please remove this line from <strong>AptaScreen.sh</strong> before using.</p>

<h2 id="usage">Usage</h2>

<p>This pipeline can only be sued for unstranded single-end sequencing.</p>

<p>This function requires the sequenced library to be as follows:                         <br />
    <strong>adapter(conserved region)-NNNNNNNNNNNNNNNNNNN-adapter(conserved regiion</strong></p>

<p>For the manual page:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>AptaScreen.sh
AptaScreen.sh --help
</code></pre></div></div>

<p>Manual page:</p>

<p><img src="/AptaScreen/images/Usage1.png" alt="" /></p>

<h2 id="examples">Examples</h2>

<p>A regular run for three screening library round #0 (R0), round #1 (R1), and round #2 (R2) containing the following reads:
     <strong>adapter(conserved region)- NNNNNNNNN (40 N) NNNNNNNNNN -adapter(conserved regiion</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code># Raw single-end sequencing data hierarchy
input/
    ├── R0.fastq.gz
    ├── R1.fastq.gz
    ├── R2.fastq.gz

# unzip the input files
gunzip -c R0.fastq.gz &gt; R0.fastq
gunzip -c R1.fastq.gz &gt; R1.fastq
gunzip -c R2.fastq.gz &gt; R2.fastq

# Run RNAseqPE
AptaScreen.sh -i R1_1000.fastq R2_1000.fastq R3_1000.fastq
              -n R0 R1 R2
              -l 40           # length of the Ns in apatmar
              -g1 GGGGGAATTCT # 5' conserved region of the apatamer
              -g2 TCTCGGATCCT # reverse complementary sequence of the 3' conserved region of the apatamer
              -t integrated_result  
              -k 100000       # number of top enriched sequences to keep
              -p 3            # number of processors 
</code></pre></div></div>

<p>After the pipeline finishes, you will get a list of outputs:</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>Outputs/
    ├── R0.fastq.csv # A table containg Aptamer sequences (unconserved region only) and counts in R0
    ├── R0.fastq.trim.processed.combined.count.fasta  # fasta file containg Aptamer sequences (unconserved region only) in R0, header of each sequence: rank-count-RPM(read per million reads)
    ├── R0.fastq.trim.processed.combined.fastq # A fastq file containg Aptamer sequences (unconserved region only). This file is not counted
    ├── R0.fastq.top100000.csv  # A table containg the top 100000 enriched Aptamer sequences (unconserved region only) and counts in R0
    ├── R0.fastq.csv.randomness.csv  # reporting the duplicate level (randomness) of R0 
    ├── R0.fastq.csv.ATGCdistribution.csv  # reporting the ATGC distribution of R0 
    ├── R0.fastq.trim.cutadpt.log  # log for cutadpt
    ├── R1.fastq.csv 
    ├── R1.fastq.trim.processed.combined.count.fasta  
    ├── R1.fastq.trim.processed.combined.fastq 
    ├── R1.fastq.top100000.csv  
    ├── R1.fastq.csv.randomness.csv  
    ├── R1.fastq.csv.ATGCdistribution.csv  
    ├── R1.fastq.trim.cutadpt.log
    ├── R2.fastq.csv 
    ├── R2.fastq.trim.processed.combined.count.fasta  
    ├── R2.fastq.trim.processed.combined.fastq 
    ├── R2.fastq.top100000.csv  
    ├── R2.fastq.csv.randomness.csv  
    ├── R2.fastq.csv.ATGCdistribution.csv  
    ├── R2.fastq.trim.cutadpt.log
    ├── Randomness.csv # reporting  duplicate levels (randomness) of all libraries (R0, R1, R2)
    ├── Randomness.pdf # A lineplot reporting  duplicate levels (randomness) of all libraries (R0, R1, R2)
    ├── Randomness_barplot.pdf # A barplot reporting  duplicate levels (randomness) of all libraries (R0, R1, R2)
    ├── Randomness_barplot.csv # data for the barplot  
    ├── reads_number_report.csv # reporting the depth of each library  
    ├── integrated_result.top100000.csv # A summerized table containing the top 100000 enriched Aptamer sequences (unconserved region only) in R2 and theirs counts in R0, R1, and R2.
    ├── R.log # R console
</code></pre></div></div>

<p>Usually, <strong>integrated_result.top100000.csv</strong> is the most useful output data.</p>


      
      <div class="footer border-top border-gray-light mt-5 pt-3 text-right text-gray">
        This site is open source. <a href="https://github.com/FengJiang94/AptaScreen/edit/main/README.md">Improve this page</a>.
      </div>
      
    </div>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/anchor-js/4.1.0/anchor.min.js" integrity="sha256-lZaRhKri35AyJSypXXs4o6OPFTbTmUoltBbDCbdzegg=" crossorigin="anonymous"></script>
    <script>anchors.add();</script>
  </body>
</html>
